### ISSUE TITLE
Unable to generate new project (ssh issue?)

### ISSUE THREAD

**Issue by signalnerve** [ashleygwilliams/cargo-generate#168]
When trying to generate a new project (on Windows 10, using WSL shell), `cargo-generate` errors out while trying to generate a project:

```
playground $ cargo generate --git https://github.com/cloudflare/worker-template
 Project Name: Testing
 Renaming project called `Testing` to `testing`...
 Creating project called `testing`...
 Git Error: failed to clone into: /mnt/c/Users/Kristian Freeman/src/playground/testing6aQq65
```

(This issue was moved from https://github.com/cloudflare/wrangler/issues/322)

**Comment by ashleygwilliams** [comment]
@steveklabnik are you able to reproduce this?

**Comment by steveklabnik** [comment]
I am not. However, I am on regular Windows, from the original report, it seems like (due to the `/mnt/c`) @signalnerve is using WSL. I'm not sure if that's relevant or not.

**Comment by signalnerve** [comment]
Yep, it is with WSL! Any other info I can provide to make diagnosing easier?

**Comment by 121watts** [comment]
This is happening on my machine as well MacOS 10.14.5 Mojave

**Comment by k0pernicus** [comment]
@121watts Can you provide us the version of `cargo-generate` you are using please, using `cargo generate -V`?

I am using too MacOS 10.14.5 and I am able to generate the same project, using your command...

Thanks in advance :)

**Comment by EverlastingBugstopper** [comment]
@k0pernicus there are a bunch of user reports here if you want to take a look https://github.com/cloudflare/wrangler/issues/316

**Comment by k0pernicus** [comment]
@EverlastingBugstopper I will take a look, thanks

**Comment by EverlastingBugstopper** [comment]
It's also important to note that all of these commands are being sent w/the `--force` flag. I see the bug occurring without the `--force` flag being present, but for the cases I can reproduce, it stops being a problem when I omit `--force`. Might be a red herring, not quite sure.

**Comment by k0pernicus** [comment]
Unfortunately, on my both Debian and macOS machines, I can't reproduce the bug...

Based on the logs, I tried to look for where can be the bug, and as I understood about this bug, I am wondering if it's not located in cargo git dependencies...
I checked on recent issues, and found some issues about hash path [here](https://github.com/rust-lang/cargo/issues/6882).

The main problem of this issue is that it's totally random...

**Comment by k0pernicus** [comment]
(Sorry, closed accidently...)

**Comment by EverlastingBugstopper** [comment]
Yeah - seems like that's the issue. The errors we see always have a random hash appended to the end of the repo name. It'd be great if we knew why the bug reproduces for some people and not others and if there's a way to fix it or work around it.

**Comment by k0pernicus** [comment]
@EverlastingBugstopper Do you know if this kind of errors is happening all the time for users that currently have this problem, or only one time over XXX?

**Comment by EverlastingBugstopper** [comment]
I believe it's every time.

**Comment by signalnerve** [comment]
Hi all - just to check back in, I'm still seeing this issue on the newest version of `cargo-generate`

**Comment by k0pernicus** [comment]
Hi @signalnerve, still on Windows? Which version of rustc & cargo are you using please?
Can you run again the command prefixed by `RUST_BACKTRACE=1`and copy/paste the output (crash) here please?
Thanks a lot! :)

**Comment by signalnerve** [comment]
> Hi all - just to check back in, I'm still seeing this issue on the newest version of `cargo-generate`

this was a _super_ unhelpful comment. sorry about that :|

I'm pretty sure that this issue is stemming from the `temp` variable in `git.rs`:

https://github.com/ashleygwilliams/cargo-generate/blob/fe16bc58ae07a4dbfb30b4506625d2257f2b33e3/src/git.rs#L45-L48

I cloned this locally and did a `println!` on both `project_dir.to_str()` _and_ `&temp.path()` - they differed, with `&temp.path()` having the extra characters that are causing me issues:

```
cargo-generate $ cargo +stable build && ./target/debug/cargo-generate generate --git https://github.com/ashleygwilliams/wasm-pack-template --name mytest
 Compiling cargo-generate v0.3.1 (/mnt/c/Users/Kristian Freeman/src/cargo-generate)
 Creating project called `mytest`...

# project_dir.to_str().unwrap()
"/mnt/c/Users/Kristian Freeman/src/cargo-generate/mytest"

# &temp.path()
"/mnt/c/Users/Kristian Freeman/src/cargo-generate/mytest9IU1Bz"

 Git Error: failed to clone into: /mnt/c/Users/Kristian Freeman/src/cargo-generate/mytest9IU1Bz
```

I don't know enough about the `tempfile` package to know if this adding the characters to the end of the path is expected behavior (I could imagine this is a way to obfuscate or indicate it's "temporary", but idk for sure) - either way, hope this is helpful!

```
cargo-generate $ rustc --version
rustc 1.36.0 (a53f9df32 2019-07-03)

cargo-generate $ cargo --version
cargo 1.36.0 (c4fcfb725 2019-05-15)
```

**Comment by k0pernicus** [comment]
I checked `tempfile` package and it seems that `tempdir` creates a temporary directory at a dedicated path, adding some random bytes at the end of the directory name (by default: 6!).

This function returns an `io::Result` type, and we check it when we create the temporary directory, so it seems the directory has been created before the `clone` process.
As the temporary directory is deleted at the end of scope, we can't know here if the temporary folder has been correctly created before to execute the `checkout` function...

I am wondering if it's not a bad behaviour of the VM as explained [here](https://github.com/Stebalien/tempfile/issues/69) , like OS permission issue, or a signal that has not been handled...

I am sorry but, as I canno't reproduce the bug, I can't help you for this issue.
The best way to find the bug is to define first precisely if the error comes from `tempfile`, the `checkout` function execution, or the `copy_to` function execution (but I doubt it).
Best way to define the location is to put a breakpoint line 46, and check line by line where the `return` is done (line 48, 51, or 54 from git.rs).
Or, you can simply put some `println!` statements after the initialization of `temp`, before **and** after the `checkout` execution function, and after the `copy_to` function execution...

**Comment by EverlastingBugstopper** [comment]
Putting this here for reference. Backtrace from a macos machine exhibiting this issue

```$ RUST_BACKTRACE=1 wrangler generate test https://github.com/nodebotanist/LED-Workers-Workshop.git
⬇️ Installing cargo-generate...
🌀  🐑  Generating a new webpack worker project with name 'test'... 🌀
🔧   Creating project called `test`...
⛔   Git Error: failed to clone into: /Users/adam/dev/projects/cf/testgpGq2I
Error: No such file or directory (os error 2)
stack backtrace:
   0: backtrace::backtrace::trace::h27a7a01d1b768196 (0x10f5cbc8e)
   1: backtrace::capture::Backtrace::new_unresolved::hc47d4bbb7be8e720 (0x10f5cad98)
   2: failure::backtrace::internal::InternalBacktrace::new::hdcbafa819eb3492b (0x10f5ca5a9)
   3: <failure::backtrace::Backtrace as core::default::Default>::default::h4c7ac0d2835dc7ae (0x10f5ca7a5)
   4: wrangler::settings::project::Project::generate::hc14a53ce7fcacdca (0x10f2d62ac)
   5: wrangler::commands::generate::generate::hf2fc924c122fe744 (0x10f29cd47)
   6: wrangler::run::h8d7b2f547df1e297 (0x10f2dc538)
   7: wrangler::main::h1798a82f4300f3cb (0x10f2da0b5)
   8: std::rt::lang_start::{{closure}}::h530d08244901b4e8 (0x10f2c80e6)
   9: std::panicking::try::do_call::h8037d9f03e27d896 (0x10f6d3638)
  10: ___rust_maybe_catch_panic (0x10f6dd06f)
  11: std::rt::lang_start_internal::hc8e69e673740d4ae (0x10f6d411e)
  12: _main (0x10f2dd349)```

**Comment by cj123** [comment]
Hi! I'm here following attempting to use Cloudflare wrangler as follows:

```
$ wrangler generate testproject https://github.com/EverlastingBugstopper/worker-typescript-template
⬇️ Installing cargo-generate...
  Generating a new webpack worker project with name 'testproject'...
 Creating project called `testproject`...
 Git Error: failed to clone into: /home/cj/Projects/testprojectVSl3HA                                
Error: No such file or directory (os error 2)
```

After a bit of googling, I found that the issue was related to my `~/.gitconfig` file.

I have the following lines (I think from some golang stuff? can't remember):

```
[url "git@github.com:"]
	insteadOf = https://github.com/
```

Commenting out those two lines fixes the above issue:

```
$ wrangler generate testproject https://github.com/EverlastingBugstopper/worker-typescript-template
⬇️ Installing cargo-generate...
  Generating a new webpack worker project with name 'testproject'...
 Creating project called `testproject`...
 Done! New project created /home/cj/Projects/testproject
```


referencing https://github.com/rust-lang/cargo/issues/3381 for pointing me in this direction.

hope this helps!

**Comment by richarddavenport** [comment]
@cj123 Thanks man! I can confirm I use ssh for git, temporarily commented out these line:
```
[url "ssh://git@github.com/"]
	insteadOf = https://github.com/
[url "ssh://git@bitbucket.org/"]
	insteadOf = https://bitbucket.org/
[url "ssh://git@gitlab.com/"]
	insteadOf = https://gitlab.com/
```

Works now.... hope it gets fixed!

**Comment by EverlastingBugstopper** [comment]
This also happens for github urls that are redirects. For instance, https://github.com/EverlastingBusgtopper/cf-typescript-boilerplate will redirect to https://github.com/EverlastingBusgtopper/worker-typescript-template, but the git implementation used by `cargo-generate` either does not follow the redirect or _does_ follow the redirect and then computes a different hash when it tries to copy the source.

```js
$ cargo generate --git https://github.com/EverlastingBusgtopper/cf-typescript-boilerplate --name test  
🔧   Creating project called `test`...
Error: ⛔   Git Error: failed to clone into: /Users/averyharnish/Documents/dev/gumdrop/testD8AW5n
```

**Comment by k0pernicus** [comment]
Hi @EverlastingBugstopper 

Do you have another example please?
I just tried to access to https://github.com/EverlastingBusgtopper/cf-typescript-boilerplate, but it seems that the repository does not exists anymore (or canno't be publicly accessed...).
Thanks in advance!

**Comment by EverlastingBugstopper** [comment]
Edited... 🤦‍♂ I manually typed "Busgtopper" instead of "Bugstopper"

**Comment by rockstar** [comment]
Just stumbled over this bug, which got me dissecting my `~/.gitconfig` on OS X. Turns out the existence of `templatedir` in my `[init]` section was interfering with the clone. Not sure why, and didn't dig into it further, but I thought I'd try and add more context to this bug to remove the seeming randomness and help track down the issue.

**Comment by shijunti19** [comment]
![image](https://user-images.githubusercontent.com/5392211/71418216-5e7d8d80-26a4-11ea-9a53-fa756eea1ba6.png)

**Comment by Jack-Works** [comment]
> @cj123 Thanks man! I can confirm I use ssh for git, temporarily commented out these line:
> 
> ```
> [url "ssh://git@github.com/"]
> 	insteadOf = https://github.com/
> [url "ssh://git@bitbucket.org/"]
> 	insteadOf = https://bitbucket.org/
> [url "ssh://git@gitlab.com/"]
> 	insteadOf = https://gitlab.com/
> ```
> 
> Works now.... hope it gets fixed!

Same issue and this fix works for me

**Comment by jcswart** [comment]
I get a similar, though slightly different error on Win10 + WSL2. Didn't want to add another issue considering how similar it is.

```
root@0a8ea053daed:/workspaces/first-wasm-mod-rust# cargo generate --git https://github.com/rustwasm/wasm-pack-template
 Project Name: rust-calculate
 Creating project called `rust-calculate`...
warning: spurious network error (2 tries remaining): failed to mmap. Could not write data: Permission denied; class=Os (2)
warning: spurious network error (1 tries remaining): failed to mmap. Could not write data: Permission denied; class=Os (2)
Error:  Git Error: failed to clone into: /workspaces/first-wasm-mod-rust/rust-calculateZmJGgm  
```
However calling git clone directly works.

```                                                           
root@0a8ea053daed:/workspaces/first-wasm-mod-rust# git clone https://github.com/rustwasm/wasm-pack-template rust-calculate
Cloning into 'rust-calculate'...
remote: Enumerating objects: 188, done.
remote: Total 188 (delta 0), reused 0 (delta 0), pack-reused 188
Receiving objects: 100% (188/188), 38.43 KiB | 855.00 KiB/s, done.
Resolving deltas: 100% (90/90), done.
```

I'm just sort of playing around with this windows setup, I'm not very familiar with windows in general.

I checked the WSL2 container's `.gitconfig` and there are no url rewrites taking place.

So is direct clone of the repo do enough? What steps could I take, using raw `git clone` as a workaround to proceed? Is `cargo generate` doing anything more that git clone would have done?

**Comment by k0pernicus** [comment]
Hi @jcswart, is it possible that your issue is related to [this one](https://github.com/rust-lang/cargo/issues/6513)?

**Comment by jayphelps** [comment]
FYI for people coming here that your gitconfig might look somewhat differently than shown in https://github.com/ashleygwilliams/cargo-generate/issues/168#issuecomment-537093190

If yours looks like this, temporarily removing or commenting out these similar lines also will workaround it:

```
# [url "git@github.com:"]
# insteadOf = https://github.com/
```

**Comment by oliverpool** [comment]
I think this issue is not limited to ssh (as analyzed by @EverlastingBugstopper ).
I commented the following lines in my .gitconfig to make it work (no ssh, but an https replacement).
```
[url "https://*******:x-oauth-basic@github.com/"]
	insteadOf = https://github.com/
```

**Comment by dfang** [comment]
commentting all url section is not enough for me.

i had to comment init section to make it work.

```
# [init]
#     templatedir = ~/.git-template-with-git-hooks
```

**Comment by xanterx** [comment]
> I get a similar, though slightly different error on Win10 + WSL2. Didn't want to add another issue considering how similar it is.
> 
> ```
> root@0a8ea053daed:/workspaces/first-wasm-mod-rust# cargo generate --git https://github.com/rustwasm/wasm-pack-template
>  Project Name: rust-calculate
>  Creating project called `rust-calculate`...
> warning: spurious network error (2 tries remaining): failed to mmap. Could not write data: Permission denied; class=Os (2)
> warning: spurious network error (1 tries remaining): failed to mmap. Could not write data: Permission denied; class=Os (2)
> Error:  Git Error: failed to clone into: /workspaces/first-wasm-mod-rust/rust-calculateZmJGgm  
> ```
> 
> However calling git clone directly works.
> 
> ```
> root@0a8ea053daed:/workspaces/first-wasm-mod-rust# git clone https://github.com/rustwasm/wasm-pack-template rust-calculate
> Cloning into 'rust-calculate'...
> remote: Enumerating objects: 188, done.
> remote: Total 188 (delta 0), reused 0 (delta 0), pack-reused 188
> Receiving objects: 100% (188/188), 38.43 KiB | 855.00 KiB/s, done.
> Resolving deltas: 100% (90/90), done.
> ```
> 
> I'm just sort of playing around with this windows setup, I'm not very familiar with windows in general.
> 
> I checked the WSL2 container's `.gitconfig` and there are no url rewrites taking place.
> 
> So is direct clone of the repo enough? What steps could I take, using raw `git clone` as a workaround to proceed? Is `cargo generate` doing anything more that git clone would have done?
> 
> **Update**: Okay so replacing the `{{project-name}}` and `{{authors}}` in the Cargo.toml and then running the build command succeeds. I ran into other node related issues and I'm giving up for now. Hopefully this helps someone else.

I am facing the same issue in Win10+WSL2, I had to execute the same command from outside of WSL (powershell/cmd) which works fine.

**Comment by ssalanski** [comment]
I also encountered this issue in Win10+WSL2, and found I was able to work around by setting the "git-fetch-with-cli" flag to true https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli
```toml
[net]
git-fetch-with-cli = true
```
seems like this bypasses the libgit library cargo otherwise uses, which [may be the issue](https://github.com/rustwasm/rustwasm.github.io/issues/52#issuecomment-498477848)

**Comment by Zageron** [comment]
Would just like to confirm that this solution also worked for me. Thank you @ssalanski !

```zsh
# Failure
cargo generate --git https://github.com/rustwasm/wasm-pack-template
 Project Name: projectName
 Creating project called `projectName`...
warning: spurious network error (2 tries remaining): failed to mmap. Could not write data: Permission denied; class=Os (2)
warning: spurious network error (1 tries remaining): failed to mmap. Could not write data: Permission denied; class=Os (2)
Error:  Git Error: failed to clone into: /some/dir

# Create or edit the cargo config
vim ~/.cargo/config.toml

# Paste the following
[net]
git-fetch-with-cli = true

# Try again
cargo generate --git https://github.com/rustwasm/wasm-pack-template
```

> I also encountered this issue in Win10+WSL2, and found I was able to work around by setting the "git-fetch-with-cli" flag to true https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli
> 
> ```toml
> [net]
> git-fetch-with-cli = true
> ```
> 
> seems like this bypasses the libgit library cargo otherwise uses, which [may be the issue](https://github.com/rustwasm/rustwasm.github.io/issues/52#issuecomment-498477848)

**Comment by k0pernicus** [comment]
Thanks to all, will close the original issue then.

**Comment by WillsB3** [comment]
I also encountered this issue on macOS Big Sur, and thought I'd post my findings just incase it's helpful to anyone else.

I found thanks to some other comments above that I did not have any `insteadOf` configuration lines in my .gitconfig, but I did have this line:

```
[init]
	templatedir = ~/.git-templates
```

Which for some reason causes the generate command to break. **When I comment out this line the generate command works as expected.**

It turned out that I don't actually have a directory at `~/.git-templates`, so perhaps that's the root cause of my problem, however I've never had an issue using `git clone` or `git init` with this line present in my config before.
